New Feature Request: #2431 Player Rating Histogram

viczommers opened the pull request:
Feature Request #2431 
## Source
- Borrowed from Lichess repo based on user feature request: https://lichess.org/stat/rating/distribution/classical
## Proposed Changes
- adds RatingsChartDistribution.tsx to Profile Section (User.tsx) below Ratings above current Game Chart
### To Discuss
- UX/UI experience? Best location/div to put in
- React component mounting raises warnings see browser log when testing
- adds chart.js to dependencies
- fetch() for /termination-api/ reading? vs get?

viczommers commented:
<img width="1192" alt="Screenshot 2025-03-22 at 06 52 40" src="https://github.com/user-attachments/assets/dafe9f7d-82cf-4068-bfa8-3dc91fefe590" />
<img width="1205" alt="Screenshot 2025-03-22 at 06 54 19" src="https://github.com/user-attachments/assets/4315efab-f8b7-450b-b521-becf2d9aa671" />

anoek commented:
Nice!

* I think it's way too big right now, how would it look to the right of the rating grid?

* On mobile the labels make the graph look pretty funky, probably need to smart scaling, or maybe it'll be a non issue if we get a good looking one that's smaller to the right of the grid:
![image](https://github.com/user-attachments/assets/50ce623e-e7f4-47c1-80f2-8b060c4d3e30)


* I think the console warning is real, that should be taken care of I think: `Failed to initialize chart: Error: Canvas is already in use. Chart with ID '0' must be destroyed before the canvas with ID 'rating_distribution_chart' can be reused.` I think that means there's some cleanup that needs to be done, possibly within the chart.js logic

* Re chart.js, I'm not too wild about adding another charting library, we have nivo being used elsewhere already - how hard of a switch would it be to use that for consistency? https://nivo.rocks/line/ has some examples and "recipes" (examples) of how to do the marker, shaded area, and labels.

* fetch() for /termination-api/ reading? vs get? `get` for consistency

viczommers commented:
- Let's try Nivo! importing a whole new module for one chart feels like overkill...
- Nivo doesn't support dual axis, but there seems to be a workaround https://github.com/plouc/nivo/issues/2666
- Labels look just fine, 45-angled x-axis ticks is a standard behaviour when there is not much horizontal width, same in Plotly, etc
- tbh I prefer tighter fits, where distribution chart is vertically-biased (bigger height-to-width ratio/ exaggerated y-axis). It makes Kurtosis / Central tendency more obvious, compared to 16:9 Desktop view where tails appear much fatter
- I feel we need to put the chart into a collapsible button inside the Ratings. that would be the least disruptive tot he current workflow

viczommers commented:
- reworked for Nivo js chart 
- added Button collapse
- added Rating-to-Rank conversion, depending on toggle
<img width="1218" alt="Screenshot 2025-03-28 at 02 47 47" src="https://github.com/user-attachments/assets/f956da9a-1efe-49fd-a1e0-4126f4b110f8" />

<img width="403" alt="Screenshot 2025-03-28 at 02 50 18" src="https://github.com/user-attachments/assets/87ebe81d-c5e4-41c2-98b5-c9880bf8cdc0" />

anoek commented:
Nice, thanks!
